squire_lib/src/ffi_rounds.rs

```diff
@@ -130,4 +133,65 @@ impl RoundId {
             }
         }
     }
+
+    /// Confirms a result
+    #[no_mangle]
+    pub unsafe extern "C" fn rid_record_result(
+        self,
+        tid: TournamentId,
+        result: RoundResult,
+        aid: AdminId,
+    ) -> bool {
+        match FFI_TOURNAMENT_REGISTRY.get().unwrap().get_mut(&tid) {
+            Some(mut tournament) => {
+                match tournament.apply_op(TournOp::AdminRecordResult(
+                    TournOfficialId::Admin(aid),
+                    RoundIdentifier::Id(self),
+                    result,
+                )) {
+                    Err(err) => {
+                        println!("[FFI]: {}", err);
+                        false
+                    }
+                    Ok(_) => true,
+                }
+            }
+            None => false,
+        }
+    }
+
+    #[no_mangle]
+    pub extern "C" fn rid_results(self, tid: TournamentId) -> *const RoundResult {
```

Comment:
This seems very unsafe. I recommend splitting this up into two functions, one for getting draws and one for getting the wins for a given player. In both cases, the return type can be a `u8`, `u32`, or whatever.

Reply:
Splitting this into two functions would be an issue because the GUI would need to lock the tournament between calls to get the entire thread safely. The reason that I did it like this was such that the results returned are atomic.

Reply:
How are you accessing the tournament currently? Are you not taking out some sort of lock to call this method?

I would rather face performance degradations rather than silent memory unsafety. Should we need to change how `RoundResult` is structured, it would almost certainly cause silent undefined behavior for the caller.

Perhaps there is a way around this. Maybe an intermediate struct with a definite layout or returning a pair of pointers would do the trick, but I would like to move any UB into a compile-time error.

Reply:
The tournament is got from the dashmap as a deep copy. Meaning that the method copies then runs any processing. The lookup uses an implicit mutex I believe.

Reply:
I think I've mentioned this before, but you don't need to deeply copy the tournament for the lookup. The DashMap handles mutual exclusion.

Reply:
the tournament is in a dash map that I retrieve from. it is possible for things to occur between getting the draws and results, which leads to race conditions. hence I want to get them all at once. the behavior is well defined.